### ISSUE TITLE
[java] New Rule: VariableCanBeInlined: Local variables should not be declared and then immediately returned or thrown

### ISSUE THREAD

**Issue by Pankraz76** [pmd/pmd#5770]
**VariableCanBeInlined**

**Proposed Category:** codestyle

**Description:**

Inline variables when they are immediately used to return or throw.

- https://sonarsource.github.io/rspec/#/rspec/S1488
- https://github.com/apache/maven/pull/2404/files#r2113612293
- https://docs.openrewrite.org/recipes/staticanalysis/inlinevariable
- https://github.com/checkstyle/checkstyle/pull/16488

**Code Sample:** This should include code, that should be flagged by the rule. If possible, the "correct" code
according to this new rule should also be demonstrated.

```java
public long ComputeDurationInMilliseconds()
{
  long duration = (((hours * 60) + minutes) * 60 + seconds ) * 1000 ;
  return duration;
}

public void DoSomething()
{
  ApplicationException myException = new ApplicationException();
  throw myException;
}
```

**Possible Properties:**

*   Should this rule be customizable via properties?

**Comment by adangel** [comment]
For the return case, PMD already has a rule: [UnnecessaryLocalBeforeReturn](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#unnecessarylocalbeforereturn)

**Comment by Pankraz76** [comment]
assuming name will be `UnnecessaryLocalBeforeThrow`.
- https://github.com/pmd/pmd/pull/5835

**Comment by oowekyala** [comment]
I think it would be better to have a single rule, maybe called `UnnecessaryLocalBeforeBranch` (I know "branch" is a bit obscure but I'm not sure what else we could call it). There is no point in having another rule that is an identical copy of the existing one.

**Comment by Pankraz76** [comment]
yes this seems considerable, but also impose breaking change.

**Comment by adangel** [comment]
> I think it would be better to have a single rule, maybe called `UnnecessaryLocalBeforeBranch` (I know "branch" is a bit obscure but I'm not sure what else we could call it). There is no point in having another rule that is an identical copy of the existing one.

What about the name `UnnecessaryLocal`? Or is this too vague (for now it would detect unnecessary local vars before return and throw, later on could detect other unnecessary locals that are used only once and could be inlined).

**Comment by Pankraz76** [comment]
> single rule

On greenfield yes. As this imposes a breaking change, it seems to be an dedicated issue.

First deliver capability, then merge/migrate.

Whats your opinion?

**Comment by adangel** [comment]
> > single rule
> 
> On greenfield yes. As this imposes a breaking change, it seems to be an dedicated issue.
> 
> First deliver capability, then merge/migrate.
> 
> Whats your opinion?

I'd keep the old rule unchanged (except for marking it deprecated) and deliver the new rule with the new capability in parallel - then everybody can migrate at their own pace.

**Comment by adangel** [comment]
@Pankraz76 Could you please summarize in this issue (consider this issue as the specification for the implementation), what has been discussed, what has been decided, etc. E.g. what is the name of the rule now? What is the category? What is the compatibility impact, what the relationship to the existing rule? Which questions are still open?
Make sure to update the issue description.

**Comment by adangel** [comment]
Possible rule names:

|Name| :+1: | :-1: |
|----|------|------|
|UnnecessaryLocalBeforeReturn (old name) |  | very specific, doesn't cover throw |
|UnnecessaryLocalBeforeBranch |  | |
|UnnecessaryLocalBeforeJump | | |
|UnnecessaryLocalBefore | | |
|UnnecessaryLocal | | maybe too unspecific? |
|UnncesssaryLocalVariable | | |
|ImmediatelyReturnedOrThrown| | |
|ImmediatelyReturnOrThrows| | |
|UnnecessaryLocalBeforeReturnOrThrows| exact description | might need another name when functionality is extended |
|VariableCanBeInlined| focus on the fix | |
| :question: | | |

Any other suggestions?

**Comment by adangel** [comment]
Right now, I'm in favor of "UnnecessaryLocalVariable" - the rule message, can then be more detailed ("because it is immediately returned" or "because it is immediately thrown" or "because it is only used once here and can be inlined").

**Comment by oowekyala** [comment]
The problem with that name imo is that it is too un specific. For instance an unused local variable is most likely unnecessary as well, but here we have two rules. OTOH for imports for instance, we folded UnusedImport and other import related rules into UnnecessaryImport. So it isn't clear why we would mean "Unnecessary" in two different meanings in those two different rules.

Maybe the difference is:
- UnnecessaryImport's fix is always to remove the import (and that was also the case for all the rules it subsumed) 
- UnusedVariable's fix could be to remove the variable declaration entirely, or just keep the initialize expression without giving it a name, name the variable something obvously unused (`unused*`) or use the variable, because it's mistakenly unused. Lots of different options. 
- the fix of the rule were talking about here would be to inline the variable.

If we categorize the rules according to their potential fixes, then while it made sense to merge the import rules, merging the variable rules would just add another case to the potential list of fixes. There is no overlap so no obvious sense to merge the rules. 

I propose that we should therefore steer away from an `UnnecessaryVariableBecauseReason` type name, and focus on the fix: since there is a single one the rule would be named `VariableCanBeInlined` or something similar.

**Comment by Pankraz76** [comment]
I would vote for whatever @oowekyala is voting, assuming have a majority then.

**Comment by adangel** [comment]
Thanks @oowekyala for your suggestion.

@Pankraz76 Then let's just use `VariableCanBeInlined`. That would be open to potential new cases. The rule category should be "codestyle", I think. Can you please update the description of this github issue?  
Also please describe, that the initial implementation of the rule should focus on suggesting inlining variables, that are immediately returned or immediately thrown. The new rule will replace the old rule [UnnecessaryLocalBeforeReturn](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#unnecessarylocalbeforereturn).  
And also mention, that - if the rule should be extended in the future to cover more cases - the example in #5771 should be considered, so that we won't suggest to refactor code into something unreadable with this rule.

**Comment by adangel** [comment]
@Pankraz76 If you want to drive this further, you need to continue working on that:

* Step one: Update this issue with the decided name (VariableCanBeInlined) and the description (which you can then reuse for the impl).
* Step two: Update your PR #5847 (code + (!) description)

**Comment by Pankraz76** [comment]
Thanks will update and finish.
